### PR TITLE
Combine functions and add singleObject type

### PR DESCRIPTION
**Pull request by JustinKuli**
There were a few places where 2 functions had a lot of logic in common, with just one or two specifics for each. For example, `handleExistsMustNotHave` and `handleMissingMustHave` both handle enforcement actions when necessary on a policy. The difference can just be described with an additional boolean `objShouldExist` in a combined function.

After updating those, some sections with if blocks and early returns with a function call could be refactored to if blocks that set some variables, then a return with a function call using those variables. For a good example, look at the `handleEnforce` function in https://github.com/open-cluster-management/config-policy-controller/commit/8dd90853eba5f9d443dd554dc7729392228ce304. I think this helps make the behavior more clear.

The new `singleObject` type addresses the situations where we were passing 7+ parameters into functions, some of which are wrapped into a `map[string]interface{}`, which is then unwrapped immediately in the function. Originally I think that was to address some code smells, but the unwrapping can lead to linting warnings. The new struct just holds some of the most commonly passed pieces, and gets the advantage of being fully typed.

These changes feel a bit like personal preference in a few places, so please comment if you have any other suggestions, or don't like it. Feedback is a gift.

### CODE REVIEW

**Comment by JustinKuli** [comment]
Reviewer note: going commit by commit in GitHub is probably helpful.

**Review by mprahl — CHANGES_REQUESTED** [review]
Nice work! Just a few comments.

File: controllers/configurationpolicy_utils.go [hunk 1]
```diff
@@ -407,15 +407,15 @@ func createResourceNameStr(names []string, namespace string, namespaced bool) (n
...
 	kind string,
 	complianceObjects map[string]map[string]interface{},
 	namespaced bool,
 	plc *policyv1.ConfigurationPolicy,
 	indx int,
 	compliant bool,
+	objShouldExist bool,
 ) (update bool) {
 	// Parse discovered resources
 	nameList := []string{}
```
[the remark this answers is not in the document]
Comment (JustinKuli) [reply controllers/configurationpolicy_utils.go:421]: Might need both... a slice is needed for the `strings.Join` and to keep a consistent ordering (iterating over a map is supposed to have effectively random ordering). But a great idea!

File: controllers/configurationpolicy_controller.go [hunk 2]
```diff
@@ -1112,77 +1112,45 @@ func getNamesOfKind(
 	return buildNameList(unstruct, complianceType, resList)
 }
 
-func handleExistsMustNotHave(
+func handleEnforce(
```
[the remark this answers is not in the document]
Comment (JustinKuli) [reply controllers/configurationpolicy_controller.go]: > it does not handle mustonlyhave enforce

The logs and previous names make it seem this way, but when I'm looking at the logic in `handleSingleObj`, I suspect it *does* handle `mustonlyhave` enforce. I'll need to poke around more, unless you already know where `mustonlyhave` is enforced? (If we find it somewhere else, then obviously you're right)

Either way, I agree that needs to be cleared up. And describing function arguments would be helpful.

File: controllers/configurationpolicy_controller.go [hunk 3]
```diff
@@ -1292,19 +1285,10 @@ func createObject(
 	return true, nil
 }
 
-func deleteObject(
-	namespaced bool, namespace, name string, rsrc schema.GroupVersionResource, dclient dynamic.Interface,
-) (deleted bool, err error) {
-	objLog := log.WithValues("name", name, "namespaced", namespaced, "namespace", namespace)
+func deleteObject(res dynamic.ResourceInterface, name string) (deleted bool, err error) {
```
[the remark this answers is not in the document]
Comment (JustinKuli) [reply controllers/configurationpolicy_controller.go:1270]: Do you mean `name` in `deleteObject`? It looks like that is the name of the object to delete...

I think there is another case where there is a `name` passed around that is like what you describe; maybe this comment is on the wrong line?

[the remark this answers is not in the document]
Comment (JustinKuli) [reply controllers/configurationpolicy_controller.go:1270]: Did you mean this to go with the `createStatus` function? The `nameStr` and `nameList` there seems to fit what you're describing

File: controllers/configurationpolicy_utils.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mprahl) [reply controllers/configurationpolicy_utils.go:421]: Good point! I would just sort the map keys just before `names := strings.Join(nameList, "; ")`.

File: controllers/configurationpolicy_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mprahl) [reply controllers/configurationpolicy_controller.go]: @JustinKuli You're right. It does handle `mustonlyhave` *only* when it does not already exist.

`checkAndUpdateResource` is handles `enforce` when the object exists but is incorrect and the compliance type is `musthave` or `mustonlyhave`.

File: controllers/configurationpolicy_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mprahl) [reply controllers/configurationpolicy_controller.go:1270]: Sorry, I misread the code.

File: controllers/configurationpolicy_controller.go [hunk 4]
```diff
@@ -1334,19 +1267,10 @@ func createObject(
...
+	objLog := log.WithValues("name", name)
```
Comment (mprahl) [inline controllers/configurationpolicy_controller.go:1271]: Could you please also log the namespace like it did previously?

File: controllers/configurationpolicy_controller.go [hunk 5]
```diff
@@ -1256,24 +1263,10 @@ func objectExists(
...
-	namespaced bool,
-	namespace string,
-	name string,
-	rsrc schema.GroupVersionResource,
-	unstruct unstructured.Unstructured,
-	dclient dynamic.Interface,
-) (created bool, err error) {
-	objLog := log.WithValues("name", name, "namespaced", namespaced, "namespace", namespace)
+func createObject(res dynamic.ResourceInterface, unstruct unstructured.Unstructured) (created bool, err error) {
+	objLog := log.WithValues("name", unstruct.GetName())
```
[the remark this answers is not in the document]
Comment (JustinKuli) [reply controllers/configurationpolicy_controller.go:1249]: Good point, it was a side-effect of moving the creation of the `dynamic.ResourceInterface` out of the function (which removed the need to pass the namespace directly).

File: controllers/configurationpolicy_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (JustinKuli) [reply controllers/configurationpolicy_controller.go]: Added a new name and a comment - let me know if that is sufficient.

File: controllers/configurationpolicy_utils.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (JustinKuli) [reply controllers/configurationpolicy_utils.go]: I like this better: they'll be actually sorted now, as opposed to whatever order they were in before - I think it was still consistent before, but still.

File: controllers/configurationpolicy_utils.go [hunk 6]
```diff
@@ -454,32 +444,44 @@ func createStatus(
...
+func sortAndJoinKeys(m map[string]bool, sep string) string {
+	keys := make([]string, len(m))
+	i := 0
+
+	for key := range m {
+		keys[i] = key
+		i++
+	}
+
+	sort.Strings(keys)
```
Comment (mprahl) [inline controllers/configurationpolicy_utils.go:482]: Optional: In case you wanted to be brave and use an insertion sort, you could do this:
https://stackoverflow.com/a/55460931

I do think it's overkill for our needs though!

File: controllers/configurationpolicy_controller.go [hunk 7]
```diff
@@ -262,11 +262,7 @@ func (r *ConfigurationPolicyReconciler) handleObjectTemplates(
 	}
 
 	// initialize the RelatedObjects for this Configuration Policy
-	oldRelated := []policyv1.RelatedObject{}
-
-	for i := range plc.Status.RelatedObjects {
-		oldRelated = append(oldRelated, plc.Status.RelatedObjects[i])
-	}
+	oldRelated := plc.Status.RelatedObjects
```
Comment (mprahl) [inline controllers/configurationpolicy_controller.go:265]: Are you sure this is safe? This is different behavior than what was previously there since even though you are assigning this to a variable, it contains the same slice `plc.Status.RelatedObjects`. The original behavior essentially created a copy.

See this playground for an example:
https://go.dev/play/p/VkDtSmWJ2ul

File: controllers/configurationpolicy_controller.go [hunk 8]
```diff
@@ -1046,8 +1042,7 @@ func getDetails(unstruct unstructured.Unstructured) (name string, kind string, n
 func buildNameList(
 	unstruct unstructured.Unstructured, complianceType string, resList *unstructured.UnstructuredList,
 ) (kindNameList []string) {
-	for i := range resList.Items {
```
Comment (mprahl) [inline controllers/configurationpolicy_controller.go:1067]: In the original code, `uObj` is the original struct but the for range I believe creates a copy. This would be totally fine except it seems that `handleSingleKey` is passed the pointer of `uObj` so is `handleSingleKey` modifying `uObj` and is the caller to `buildNameList` expecting its argument of `resList` to be modified?

File: controllers/configurationpolicy_controller.go (hunk 7 above)
Comment (JustinKuli) [reply controllers/configurationpolicy_controller.go:265]: Yeah, you're right, this is a rough edge I've run into before - slices are pointers to backing arrays, so you can get into weird stuff...

The tests passed, which made me more confident, but you're right that it could still run into trouble.

Comment (mprahl) [reply controllers/configurationpolicy_controller.go:265]: I'm good with it if the Config Policy controller e2e tests from the policy framework also pass.

File: controllers/configurationpolicy_controller.go (hunk 8 above)
Comment (JustinKuli) [reply controllers/configurationpolicy_controller.go:1067]: These comments about pointers and copies are giving me flashbacks to my Advent of Code stuff with Rust 😓 

Just looking at the code around this change, I think you are correct in general, the `range` makes a copy, which means I may have altered the behavior. I think in another branch of refactoring that I threw away, I changed this to not be a pointer at all and it still worked. So I don't think we use the mutability of the original thing (as in, the copy is fine). BUT making an unnecessary copy of an entire `unstruct` could be getting into the realm where passing a pointer is better for performance.

tl;dr I will revert 😆

**Review by mprahl — APPROVED** [review]
Great job! Thanks for doing this.

/hold for rebase and merge strategy and perhaps another reviewer
